// resium/dist/resium.js
import Le, { createContext as _e, useContext as Oe, useRef as E, useEffect as Y, useState as Te, useCallback as ne, useLayoutEffect as Ae, useImperativeHandle as Be, forwardRef as Me, useMemo as Ge } from "react";
import { ScreenSpaceEventType as v, ScreenSpaceEventHandler as ke, Event as he, BillboardCollection as Ie, BillboardGraphics as Fe, BoxGraphics as Ve, Cesium3DTileset as Ue, Cesium3DTilesetGraphics as We, Cesium3DTilesTerrainProvider as fe, CesiumWidget as He, ClassificationPrimitive as je, CloudCollection as Ne, CorridorGraphics as ze, CustomDataSource as xe, CylinderGraphics as qe, CzmlDataSource as Ye, EllipseGraphics as Je, EllipsoidGraphics as Ke, Entity as Ze, Fog as ye, GeoJsonDataSource as Xe, Google2DImageryProvider as pe, GroundPolylinePrimitive as Qe, GroundPrimitive as ei, ImageryLayer as ii, KmlDataSource as ti, LabelCollection as ni, LabelGraphics as oi, Model as ri, ModelGraphics as si, Moon as ge, ParticleSystem as ai, PathGraphics as ci, PlaneGraphics as li, PointGraphics as ui, PointPrimitiveCollection as di, PolygonGraphics as mi, PolylineCollection as hi, PolylineGraphics as fi, PolylineVolumeGraphics as yi, PostProcessStage as pi, PostProcessStageLibrary as F, PostProcessStageComposite as gi, Primitive as vi, RectangleGraphics as Ci, SceneMode as le, Sun as ve, TimeDynamicPointCloud as Pi, Viewer as Ei, WallGraphics as wi } from "cesium";
import { createPortal as Si } from "react-dom";
const me = _e({}), { Provider: Kn, Consumer: bi } = me, re = () => Oe(me) || {}, se = (i, e) => {
  const t = (n) => {
    const r = re(), c = E(void 0), l = E(!1);
    return Y(() => () => {
      r.camera && n.cancelFlightOnUnmount && r.camera.cancelFlight();
    }, [r.camera, n.cancelFlightOnUnmount]), Y(() => {
      r.camera && r.scene && !r.scene.isDestroyed() && (!n.once || !l.current) && (r.camera.cancelFlight(), e(r.camera, n, c.current), l.current = !0), c.current = n;
    }), null;
  };
  return t.displayName = i, t;
};
var oe = { exports: {} }, Q = {};
var Ce;
function Di() {
  if (Ce) return Q;
  Ce = 1;
  var i = /* @__PURE__ */ Symbol.for("react.transitional.element"), e = /* @__PURE__ */ Symbol.for("react.fragment");
  function t(n, r, c) {
    var l = null;
    if (c !== void 0 && (l = "" + c), r.key !== void 0 && (l = "" + r.key), "key" in r) {
      c = {};
      for (var m in r)
        m !== "key" && (c[m] = r[m]);
    } else c = r;
    return r = c.ref, {
      $$typeof: i,
      type: n,
      key: l,
      ref: r !== void 0 ? r : null,
      props: c
    };
  }
  return Q.Fragment = e, Q.jsx = t, Q.jsxs = t, Q;
}
var ee = {};
var Pe;
function Ri() {
  return Pe || (Pe = 1, process.env.NODE_ENV !== "production" && (function() {
    function i(o) {
      if (o == null) return null;
      if (typeof o == "function")
        return o.$$typeof === Z ? null : o.displayName || o.name || null;
      if (typeof o == "string") return o;
      switch (o) {
        case L:
          return "Fragment";
        case k:
          return "Profiler";
        case M:
          return "StrictMode";
        case J:
          return "Suspense";
        case N:
          return "SuspenseList";
        case U:
          return "Activity";
      }
      if (typeof o == "object")
        switch (typeof o.tag == "number" && console.error(
          "Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."
        ), o.$$typeof) {
          case B:
            return "Portal";
          case te:
            return o.displayName || "Context";
          case V:
            return (o._context.displayName || "Context") + ".Consumer";
          case j:
            var s = o.render;
            return o = o.displayName, o || (o = s.displayName || s.name || "", o = o !== "" ? "ForwardRef(" + o + ")" : "ForwardRef"), o;
          case K:
            return s = o.displayName || null, s !== null ? s : i(o.type) || "Memo";
          case D:
            s = o._payload, o = o._init;
            try {
              return i(o(s));
            } catch {
            }
        }
      return null;
    }
    function e(o) {
      return "" + o;
    }
    function t(o) {
      try {
        e(o);
        var s = !1;
      } catch {
        s = !0;
      }
      if (s) {
        s = console;
        var h = s.error, f = typeof Symbol == "function" && Symbol.toStringTag && o[Symbol.toStringTag] || o.constructor.name || "Object";
        return h.call(
          s,
          "The provided key is an unsupported type %s. This value must be coerced to a string before using it here.",
          f
        ), e(o);
      }
    }
    function n(o) {
      if (o === L) return "<>";
      if (typeof o == "object" && o !== null && o.$$typeof === D)
        return "<...>";
      try {
        var s = i(o);
        return s ? "<" + s + ">" : "<...>";
      } catch {
        return "<...>";
      }
    }
    function r() {
      var o = W.A;
      return o === null ? null : o.getOwner();
    }
    function c() {
      return Error("react-stack-top-frame");
    }
    function l(o) {
      if (z.call(o, "key")) {
        var s = Object.getOwnPropertyDescriptor(o, "key").get;
        if (s && s.isReactWarning) return !1;
      }
      return o.key !== void 0;
    }
    function m(o, s) {
      function h() {
        p || (p = !0, console.error(
          "%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://react.dev/link/special-props)",
          s
        ));
      }
      h.isReactWarning = !0, Object.defineProperty(o, "key", {
        get: h,
        configurable: !0
      });
    }
    function u() {
      var o = i(this.type);
      return R[o] || (R[o] = !0, console.error(
        "Accessing element.ref was removed in React 19. ref is now a regular prop. It will be removed from the JSX Element type in a future release."
      )), o = this.props.ref, o !== void 0 ? o : null;
    }
    function y(o, s, h, f, G, A) {
      var C = h.ref;
      return o = {
        $$typeof: b,
        type: o,
        key: s,
        props: h,
        _owner: f
      }, (C !== void 0 ? C : null) !== null ? Object.defineProperty(o, "ref", {
        enumerable: !1,
        get: u
      }) : Object.defineProperty(o, "ref", { enumerable: !1, value: null }), o._store = {}, Object.defineProperty(o._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: 0
      }), Object.defineProperty(o, "_debugInfo", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: null
      }), Object.defineProperty(o, "_debugStack", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: G
      }), Object.defineProperty(o, "_debugTask", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: A
      }), Object.freeze && (Object.freeze(o.props), Object.freeze(o)), o;
    }
    function P(o, s, h, f, G, A) {
      var C = s.children;
      if (C !== void 0)
        if (f)
          if (X(C)) {
            for (f = 0; f < C.length; f++)
              w(C[f]);
            Object.freeze && Object.freeze(C);
          } else
            console.error(
              "React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead."
            );
        else w(C);
      if (z.call(s, "key")) {
        C = i(o);
        var x = Object.keys(s).filter(function($e) {
          return $e !== "key";
        });
        f = 0 < x.length ? "{key: someKey, " + x.join(": ..., ") + ": ...}" : "{key: someKey}", O[C + f] || (x = 0 < x.length ? "{" + x.join(": ..., ") + ": ...}" : "{}", console.error(
          `A props object containing a "key" prop is being spread into JSX:
  let props = %s;
  <%s {...props} />
React keys must be passed directly to JSX without using spread:
  let props = %s;
  <%s key={someKey} {...props} />`,
          f,
          C,
          x,
          C
        ), O[C + f] = !0);
      }
      if (C = null, h !== void 0 && (t(h), C = "" + h), l(s) && (t(s.key), C = "" + s.key), "key" in s) {
        h = {};
        for (var ce in s)
          ce !== "key" && (h[ce] = s[ce]);
      } else h = s;
      return C && m(
        h,
        typeof o == "function" ? o.displayName || o.name || "Unknown" : o
      ), y(
        o,
        C,
        h,
        r(),
        G,
        A
      );
    }
    function w(o) {
      S(o) ? o._store && (o._store.validated = 1) : typeof o == "object" && o !== null && o.$$typeof === D && (o._payload.status === "fulfilled" ? S(o._payload.value) && o._payload.value._store && (o._payload.value._store.validated = 1) : o._store && (o._store.validated = 1));
    }
    function S(o) {
      return typeof o == "object" && o !== null && o.$$typeof === b;
    }
    var d = Le, b = /* @__PURE__ */ Symbol.for("react.transitional.element"), B = /* @__PURE__ */ Symbol.for("react.portal"), L = /* @__PURE__ */ Symbol.for("react.fragment"), M = /* @__PURE__ */ Symbol.for("react.strict_mode"), k = /* @__PURE__ */ Symbol.for("react.profiler"), V = /* @__PURE__ */ Symbol.for("react.consumer"), te = /* @__PURE__ */ Symbol.for("react.context"), j = /* @__PURE__ */ Symbol.for("react.forward_ref"), J = /* @__PURE__ */ Symbol.for("react.suspense"), N = /* @__PURE__ */ Symbol.for("react.suspense_list"), K = /* @__PURE__ */ Symbol.for("react.memo"), D = /* @__PURE__ */ Symbol.for("react.lazy"), U = /* @__PURE__ */ Symbol.for("react.activity"), Z = /* @__PURE__ */ Symbol.for("react.client.reference"), W = d.__CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE, z = Object.prototype.hasOwnProperty, X = Array.isArray, g = console.createTask ? console.createTask : function() {
      return null;
    };
    d = {
      react_stack_bottom_frame: function(o) {
        return o();
      }
    };
    var p, R = {}, _ = d.react_stack_bottom_frame.bind(
      d,
      c
    )(), H = g(n(c)), O = {};
    ee.Fragment = L, ee.jsx = function(o, s, h) {
      var f = 1e4 > W.recentlyCreatedOwnerStacks++;
      return P(
        o,
        s,
        h,
        !1,
        f ? Error("react-stack-top-frame") : _,
        f ? g(n(o)) : H
      );
    }, ee.jsxs = function(o, s, h) {
      var f = 1e4 > W.recentlyCreatedOwnerStacks++;
      return P(
        o,
        s,
        h,
        !0,
        f ? Error("react-stack-top-frame") : _,
        f ? g(n(o)) : H
      );
    };
  })()), ee;
}
var Ee;
function Ti() {
  return Ee || (Ee = 1, process.env.NODE_ENV === "production" ? oe.exports = Di() : oe.exports = Ri()), oe.exports;
}
var $ = Ti();
function Mi(i, e) {
  return e ? ue(i).reduce(
    (t, [n, r]) => (ie(e, n) && (t[n] = r), t),
    {}
  ) : {};
}
function ue(i) {
  return Object.keys(i).map((e) => [e, i[e]]);
}
function ie(i, e) {
  return !!i && i.indexOf(e) !== -1;
}
function ki(i, e) {
  return !!i && !!e && [...Object.keys(i), ...Object.keys(e)].every((t) => i[t] === e[t]);
}
function $i(i) {
  return i && typeof i.isDestroyed == "function" && typeof i.destroy == "function";
}
function Li(i) {
  return $i(i) && i.isDestroyed();
}
function I(i) {
  return i && typeof i == "object" && "then" in i && typeof i.then == "function";
}
const q = "__RESIUM_EVENT_MANAGER", de = [
  "onClick",
  "onDoubleClick",
  "onMouseDown",
  "onMouseUp",
  "onMiddleClick",
  "onMiddleDown",
  "onMiddleUp",
  "onMouseMove",
  "onPinchEnd",
  "onPinchMove",
  "onPinchStart",
  "onRightClick",
  "onRightDown",
  "onRightUp",
  "onWheel",
  "onMouseEnter",
  "onMouseLeave"
];
class ae {
  static eventTypeMap = {
    onClick: v.LEFT_CLICK,
    onDoubleClick: v.LEFT_DOUBLE_CLICK,
    onMouseDown: v.LEFT_DOWN,
    onMouseUp: v.LEFT_UP,
    onMiddleClick: v.MIDDLE_CLICK,
    onMiddleDown: v.MIDDLE_DOWN,
    onMiddleUp: v.MIDDLE_UP,
    onMouseMove: v.MOUSE_MOVE,
    onPinchEnd: v.PINCH_END,
    onPinchMove: v.PINCH_MOVE,
    onPinchStart: v.PINCH_START,
    onRightClick: v.RIGHT_CLICK,
    onRightDown: v.RIGHT_DOWN,
    onRightUp: v.RIGHT_UP,
    onWheel: v.WHEEL,
    onMouseEnter: v.MOUSE_MOVE,
    onMouseLeave: v.MOUSE_MOVE
  };
  scene;
  sshe;
  events = {
    onClick: /* @__PURE__ */ new Map(),
    onDoubleClick: /* @__PURE__ */ new Map(),
    onMouseDown: /* @__PURE__ */ new Map(),
    onMouseUp: /* @__PURE__ */ new Map(),
    onMiddleClick: /* @__PURE__ */ new Map(),
    onMiddleDown: /* @__PURE__ */ new Map(),
    onMiddleUp: /* @__PURE__ */ new Map(),
    onMouseMove: /* @__PURE__ */ new Map(),
    onPinchEnd: /* @__PURE__ */ new Map(),
    onPinchMove: /* @__PURE__ */ new Map(),
    onPinchStart: /* @__PURE__ */ new Map(),
    onRightClick: /* @__PURE__ */ new Map(),
    onRightDown: /* @__PURE__ */ new Map(),
    onRightUp: /* @__PURE__ */ new Map(),
    onWheel: /* @__PURE__ */ new Map(),
    onMouseEnter: /* @__PURE__ */ new Map(),
    onMouseLeave: /* @__PURE__ */ new Map()
  };
  hovered = void 0;
  constructor(e) {
    this.scene = e, this.sshe = new ke(e?.canvas);
  }
  destroy() {
    this.hovered = void 0, this.sshe.isDestroyed() || this.sshe.destroy();
  }
  isDestroyed() {
    return this.sshe.isDestroyed();
  }
  on(e, t, n) {
    e && t === "onWheel" || this.events[t].set(e, n);
  }
  off(e, t) {
    this.events[t].delete(e), this.hovered === e && (this.hovered = void 0);
  }
  setEvents(e, t) {
    ue(t).forEach(([n, r]) => {
      const c = n;
      ie(de, c) && (r ? this.on(e, c, r) : this.off(e, c));
    }), this.commit();
  }
  clearEvents(e) {
    this.hovered = void 0, de.forEach((t) => {
      this.off(e, t);
    }), this.commit();
  }
  commit() {
    const e = this.sshe, t = this.sshe.isDestroyed();
    t || (this.events.onMouseEnter.size === 0 && this.events.onMouseLeave.size === 0 && this.events.onMouseMove.size === 0 ? this.sshe.removeInputAction(v.MOUSE_MOVE) : this.sshe.getInputAction(v.MOUSE_MOVE) || this.sshe.setInputAction(this.onMouseMove, v.MOUSE_MOVE)), ue(this.events).forEach(([n, r]) => {
      if (n === "onMouseEnter" || n === "onMouseLeave" || n === "onMouseMove")
        return;
      const c = ae.eventTypeMap[n];
      t || (r.size === 0 ? e.removeInputAction(c) : e.getInputAction(c) || e.setInputAction(this.eventCallback(n), c));
    });
  }
  getScreenSpaceEventHandler() {
    return this.sshe;
  }
  getEventCallback(e, t) {
    return t === null ? this.events[e].get(null) : this.events[e].get(t.id) || // Entity
    this.events[e].get(t.id?.entityCollection?.owner) || // Entity in DataSource
    this.events[e].get(t.primitive) || // Primitive
    this.events[e].get(t.tileset);
  }
  onMouseMove = (e) => {
    const t = this.pick(e.endPosition);
    this.hovered !== t && (this.hovered && (this.getEventCallback("onMouseLeave", this.hovered)?.(e, this.hovered), this.getEventCallback("onMouseLeave", null)?.(e, this.hovered)), t && (this.getEventCallback("onMouseEnter", t)?.(e, t), this.getEventCallback("onMouseEnter", null)?.(e, t))), t && this.getEventCallback("onMouseMove", t)?.(e, t), this.getEventCallback("onMouseMove", null)?.(e, t), this.hovered = t;
  };
  eventCallback = (e) => (t) => {
    const n = this.pick(t?.position);
    n && this.getEventCallback(e, n)?.(t, n), this.getEventCallback(e, null)?.(t, n);
  };
  pick(e) {
    if (e)
      return this.scene?.pick(e);
  }
}
const _i = ({
  name: i,
  create: e,
  destroy: t,
  provide: n,
  update: r,
  cesiumReadonlyProps: c,
  cesiumEventProps: l,
  otherProps: m,
  setCesiumPropsAfterCreate: u,
  useCommonEvent: y,
  useRootEvent: P
}, w, S) => {
  const d = E(void 0), b = re(), B = E(n ? {} : void 0), L = E({}), M = E(we(w)), k = E({}), [V, te] = Te(!1), j = E(!1), J = E(null), N = E(void 0), K = b?.[q], D = E(void 0), U = E(void 0), Z = ne(
    async (g) => {
      if (!d.current) return;
      const p = d.current, R = Object.keys(g), _ = Object.keys(l || []), H = R.concat(
        Object.keys(k.current).filter((s) => !R.includes(s))
      ).filter((s) => k.current[s] !== g[s]).map((s) => [s, k.current[s], g[s]]), O = [];
      for (const [s, h, f] of H)
        if (c?.includes(s))
          O.push(s);
        else if (ie(_, s)) {
          const G = l?.[s], A = p[G];
          A instanceof he && (typeof h > "u" ? (A.addEventListener(f), L.current[G] = f) : typeof f > "u" ? (A.removeEventListener(h), delete L.current[G]) : (A.removeEventListener(h), A.addEventListener(f)));
        } else s !== "children" && !de.includes(s) && !m?.includes(s) && (p[s] = f);
      const o = P ? B.current?.[q] : K;
      if (y && o && d.current && o.setEvents(P ? null : d.current, g), r && j.current) {
        const s = r(d.current, g, k.current, b);
        I(s) && await s;
      }
      k.current = g, M.current = g, j.current && O.length > 0 && (process.env.NODE_ENV !== "production" && console.warn(
        `Warning: <${i}> is recreated because following read-only props have been updated: ${O.join(
          ", "
        )}`
      ), z(), await X(), D.current = W());
    },
    []
    // eslint-disable-line react-hooks/exhaustive-deps
  ), W = ne(async () => {
    await new Promise((_) => queueMicrotask(() => _(void 0)));
    const g = e?.(b, M.current, J.current);
    let p;
    if (I(g) ? p = await g : p = g, Array.isArray(p) ? (d.current = p[0], N.current = p[1]) : d.current = p, u)
      await Z(M.current);
    else {
      if (d.current && l) {
        const _ = d.current;
        for (const H of Object.keys(M.current)) {
          const O = l[H];
          if (O) {
            const o = M.current[H], s = _[O];
            o && s instanceof he && s.addEventListener(o);
          }
        }
      }
      k.current = M.current;
    }
    n && d.current && (B.current = {
      ...b,
      ...n(d.current, b, w, N.current)
    });
    const R = P ? B.current?.[q] : K;
    y && R && d.current && R.setEvents(P ? null : d.current, M.current), U.current || te(!0);
  }, []), z = ne(() => {
    te(!1), j.current = !1;
  }, []), X = ne(async () => {
    await new Promise((p) => queueMicrotask(() => p(void 0))), D.current && (await D.current, D.current = void 0), d.current && t && t(d.current, b, J.current, N.current);
    const g = P ? B.current?.[q] : K;
    if (y && g && d.current && g.clearEvents(P ? null : d.current), d.current && !Li(d.current)) {
      const p = Object.keys(L.current);
      for (const R of p)
        d.current[R]?.removeEventListener?.(L.current[R]);
    }
    L.current = {}, B.current = void 0, N.current = void 0, d.current = void 0;
  }, []);
  return Ae(() => ((async () => (U.current && (await U.current, U.current = void 0), D.current = W()))(), () => {
    z(), U.current = X();
  }), [W, X, z]), Y(() => {
    (async () => {
      D.current && await D.current;
      const p = we(w);
      V ? ki(p, k.current) || (await Z(p), b.__$internal?.onUpdate?.()) : (k.current = p, M.current = p, j.current = !0);
    })();
  }, [b.__$internal, V, w, Z]), Be(
    S,
    () => ({
      cesiumElement: V ? d.current : null
    }),
    [V]
  ), [B.current, V, J];
};
function we(i) {
  const { children: e, ...t } = i;
  return t;
}
const a = ({
  renderContainer: i,
  noChildren: e,
  containerProps: t,
  defaultProps: n,
  ...r
}) => {
  const c = (l, m) => {
    const u = {
      ...n,
      ...l
    }, [y, P, w] = _i(
      r,
      u,
      m
    );
    if (e) return null;
    const S = P && "children" in u ? u.children : null, d = i ? /* @__PURE__ */ $.jsx(
      "div",
      {
        "data-testid": "resium-container",
        ref: w,
        ...typeof t == "function" ? t(u) : Mi(u, t),
        children: S
      }
    ) : S ? /* @__PURE__ */ $.jsx($.Fragment, { children: S }) : null;
    return y ? /* @__PURE__ */ $.jsx(me.Provider, { value: y, children: d }) : d;
  };
  return c.displayName = r.name, Me(c);
}, Oi = ["enabled", "selected"], T = (i) => a({
  name: i.name,
  create(e, t) {
    if (!e.scene) return;
    const n = i.create(t, e.scene.postProcessStages);
    return typeof t.enabled == "boolean" && (n.enabled = t.enabled), t.selected && "selected" in n && (n.selected = t.selected), i.props.forEach((r) => {
      !ie(i.readonlyProps, r) && typeof t[r] < "u" && (n.uniforms[r] = t[r]);
    }), !i.noMount && e.scene && !e.scene.isDestroyed() && e.scene.postProcessStages.add(n), n;
  },
  destroy(e, t) {
    i.noMount ? e.enabled = !1 : (t.scene && !t.scene.isDestroyed() && t.scene.postProcessStages.remove(e), e.isDestroyed() || e.destroy());
  },
  update(e, t, n) {
    i.props.forEach((r) => {
      !ie(i.readonlyProps, r) && t[r] !== n[r] && (e.uniforms[r] = t[r]);
    });
  },
  cesiumProps: Oi,
  cesiumReadonlyProps: i.readonlyProps,
  defaultProps: {
    enabled: !0
  }
}), Zn = (i) => (
  // supports both functional components and class components
  Me((e, t) => /* @__PURE__ */ $.jsx(bi, { children: (n) => /* @__PURE__ */ $.jsx(i, { ...e, ref: t, cesium: n }) }))
), Ai = [
  "alignedAxis",
  "color",
  "disableDepthTestDistance",
  "distanceDisplayCondition",
  "eyeOffset",
  "height",
  "heightReference",
  "horizontalOrigin",
  "image",
  "pixelOffset",
  "pixelOffsetScaleByDistance",
  "position",
  "rotation",
  "scale",
  "scaleByDistance",
  "show",
  "sizeInMeters",
  "splitDirection",
  "translucencyByDistance",
  "verticalOrigin",
  "width",
  "id"
], Xn = a({
  name: "Billboard",
  create(i, e) {
    return i.billboardCollection?.add(e);
  },
  destroy(i, e) {
    e.billboardCollection && !e.billboardCollection.isDestroyed() && e.billboardCollection.remove(i);
  },
  cesiumProps: Ai,
  useCommonEvent: !0
}), Bi = [
  "blendOption",
  "debugShowBoundingVolume",
  "debugShowTextureAtlas",
  "modelMatrix",
  "show"
], Qn = a({
  name: "BillboardCollection",
  create(i, e) {
    if (!i.primitiveCollection) return;
    const t = new Ie({
      modelMatrix: e.modelMatrix,
      debugShowBoundingVolume: e.debugShowBoundingVolume,
      scene: i.scene,
      blendOption: e.blendOption
    });
    return i.primitiveCollection.add(t), t;
  },
  destroy(i, e) {
    e.primitiveCollection && !e.primitiveCollection.isDestroyed() && e.primitiveCollection.remove(i), i.isDestroyed() || i.destroy();
  },
  provide(i) {
    return {
      billboardCollection: i
    };
  },
  cesiumProps: Bi
}), Gi = [
  "image",
  "show",
  "scale",
  "horizontalOrigin",
  "verticalOrigin",
  "eyeOffset",
  "pixelOffset",
  "rotation",
  "alignedAxis",
  "width",
  "height",
  "color",
  "scaleByDistance",
  "translucencyByDistance",
  "pixelOffsetScaleByDistance",
  "imageSubRegion",
  "sizeInMeters",
  "heightReference",
  "distanceDisplayCondition",
  "disableDepthTestDistance",
  "splitDirection"
], Ii = {
  onDefinitionChange: "definitionChanged"
}, eo = a({
  name: "BillboardGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new Fe(e);
    return i.entity.billboard = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.billboard = void 0);
  },
  cesiumProps: Gi,
  cesiumEventProps: Ii
}), Fi = [
  "heightReference",
  "dimensions",
  "show",
  "fill",
  "material",
  "outline",
  "outlineColor",
  "outlineWidth",
  "shadows",
  "distanceDisplayCondition"
], Vi = {
  onDefinitionChange: "definitionChanged"
}, io = a({
  name: "BoxGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new Ve(e);
    return i.entity.box = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.box = void 0);
  },
  cesiumProps: Fi,
  cesiumEventProps: Vi
}), Ui = [
  "position",
  "direction",
  "up",
  "right",
  "frustum",
  "defaultMoveAmount",
  "defaultLookAmount",
  "defaultRotateAmount",
  "defaultZoomAmount",
  "constrainedAxis",
  "maximumZoomFactor",
  "percentageChanged"
], Wi = {
  onChange: "changed",
  onMoveEnd: "moveEnd",
  onMoveStart: "moveStart"
}, to = a({
  name: "Camera",
  create: (i) => i.scene?.camera,
  cesiumProps: Ui,
  cesiumEventProps: Wi,
  setCesiumPropsAfterCreate: !0
}), no = se(
  "CameraFlyHome",
  (i, { duration: e }) => {
    i.flyHome(e);
  }
), oo = se(
  "CameraFlyTo",
  (i, { onComplete: e, onCancel: t, ...n }) => {
    i.flyTo({ ...n, complete: e, cancel: t });
  }
), ro = se(
  "CameraLookAt",
  (i, { target: e, offset: t }) => {
    i.lookAt(e, t);
  }
), so = se(
  "CameraFlyToBoundingSphere",
  (i, { boundingSphere: e, onComplete: t, onCancel: n, ...r }) => {
    i.flyToBoundingSphere(e, {
      ...r,
      complete: t,
      cancel: n
    });
  }
), Hi = [
  "show",
  "modelMatrix",
  "shadows",
  "maximumScreenSpaceError",
  "cullRequestsWhileMoving",
  "cullRequestsWhileMovingMultiplier",
  "preloadWhenHidden",
  "preloadFlightDestinations",
  "preferLeaves",
  "progressiveResolutionHeightFraction",
  "foveatedScreenSpaceError",
  "foveatedConeSize",
  "foveatedMinimumScreenSpaceErrorRelaxation",
  "foveatedInterpolationCallback",
  "foveatedTimeDelay",
  "dynamicScreenSpaceError",
  "dynamicScreenSpaceErrorDensity",
  "dynamicScreenSpaceErrorFactor",
  "dynamicScreenSpaceErrorHeightFalloff",
  "skipLevelOfDetail",
  "baseScreenSpaceError",
  "skipScreenSpaceErrorFactor",
  "skipLevels",
  "immediatelyLoadDesiredLevelOfDetail",
  "loadSiblings",
  "clippingPlanes",
  "clippingPolygons",
  "lightColor",
  "colorBlendAmount",
  "colorBlendMode",
  "debugFreezeFrame",
  "debugColorizeTiles",
  "debugWireframe",
  "debugShowBoundingVolume",
  "debugShowContentBoundingVolume",
  "debugShowViewerRequestVolume",
  "debugShowGeometricError",
  "debugShowRenderingStatistics",
  "debugShowMemoryUsage",
  "debugShowUrl",
  "style",
  "backFaceCulling",
  "showOutline",
  "vectorClassificationOnly",
  "vectorKeepDecodedPositions",
  "splitDirection",
  "customShader",
  "imageBasedLighting",
  "showCreditsOnScreen",
  "featureIdLabel",
  "instanceFeatureIdLabel",
  "outlineColor",
  "cacheBytes",
  "maximumCacheOverflowBytes",
  "enableCollision"
], ji = [
  "asynchronouslyLoadImagery",
  "classificationType",
  "cullWithChildrenBounds",
  "debugHeatmapTilePropertyName",
  "ellipsoid",
  "enableDebugWireframe",
  "heightReference",
  "modelUpAxis",
  "modelForwardAxis",
  "projectTo2D",
  "enableShowOutline",
  "enablePick",
  "environmentMapOptions",
  "scene"
], Ni = {
  onAllTilesLoad: "allTilesLoaded",
  onInitialTilesLoad: "initialTilesLoaded",
  onLoadProgress: "loadProgress",
  onTileFailed: "tileFailed",
  onTileLoad: "tileLoad",
  onTileUnload: "tileUnload",
  onTileVisible: "tileVisible"
}, zi = ["onReady", "onError", "url"], ao = a({
  name: "Cesium3DTileset",
  async create(i, e) {
    if (!i.primitiveCollection) return;
    const t = e.url;
    let n;
    I(t) ? n = await t : n = t;
    let r;
    try {
      r = await Ue.fromUrl(n, e), e.onReady?.(r);
    } catch (c) {
      e.onError?.(c);
      return;
    }
    return e.colorBlendAmount && (r.colorBlendAmount = e.colorBlendAmount), e.colorBlendMode && (r.colorBlendMode = e.colorBlendMode), e.style && (r.style = e.style), i.primitiveCollection.add(r), r;
  },
  destroy(i, e) {
    e.primitiveCollection && !e.primitiveCollection.isDestroyed() && e.primitiveCollection.remove(i), i.isDestroyed() || i.destroy();
  },
  cesiumProps: Hi,
  cesiumReadonlyProps: ji,
  cesiumEventProps: Ni,
  otherProps: zi,
  useCommonEvent: !0
}), xi = ["show", "uri", "maximumScreenSpaceError"], qi = {
  onDefinitionChange: "definitionChanged"
}, co = a({
  name: "Cesium3DTilesetGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new We(e);
    return i.entity.tileset = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.tileset = void 0);
  },
  cesiumProps: xi,
  cesiumEventProps: qi
}), lo = a({
  name: "Cesium3DTilesTerrainProvider",
  async create(i, e) {
    const { onReady: t, assetId: n, accessToken: r, url: c, children: l, ...m } = e;
    let u;
    if (n !== void 0)
      u = await fe.fromIonAssetId(
        n,
        m
      );
    else if (c !== void 0)
      u = await fe.fromUrl(c, m);
    else
      throw new Error(
        "Cesium3DTilesTerrainProvider requires either 'url' or 'assetId' prop to be provided"
      );
    return t && t(u), u;
  }
}), Yi = [
  "resolutionScale",
  "useDefaultRenderLoop",
  "targetFrameRate",
  "useBrowserRecommendedResolution",
  "allowDataSourcesToSuspendAnimation",
  "trackedEntity",
  "clockTrackedDataSource",
  "terrainProvider",
  "creditDisplay"
], Ji = [
  "clock",
  "shouldAnimate",
  "ellipsoid",
  "baseLayer",
  "skyBox",
  "skyAtmosphere",
  "sceneMode",
  "scene3DOnly",
  "orderIndependentTranslucency",
  "mapMode2D",
  "mapProjection",
  "globe",
  "showRenderLoopErrors",
  "automaticallyTrackDataSourceClocks",
  "contextOptions",
  "creditContainer",
  "creditViewport",
  "dataSources",
  "shadows",
  "terrainShadows",
  "terrain",
  "requestRenderMode",
  "maximumRenderTimeChange",
  "msaaSamples",
  "blurActiveElementOnCanvasFocus"
], Ki = ["className", "id", "style", "full", "containerProps"], uo = a({
  name: "CesiumWidget",
  create(i, e, t) {
    if (!t) return;
    const n = new He(t, e);
    if (!n) return;
    typeof e.resolutionScale == "number" && (n.resolutionScale = e.resolutionScale);
    const r = new ae(n.scene);
    return [n, r];
  },
  destroy(i, e, t, n) {
    n && !n.isDestroyed() && n.destroy(), i.isDestroyed() || i.destroy();
  },
  provide(i, e, t, n) {
    return {
      cesiumWidget: i,
      scene: i.scene,
      camera: i.scene.camera,
      imageryLayerCollection: i.scene.globe.imageryLayers,
      primitiveCollection: i.scene.primitives,
      globe: i.scene.globe,
      __$internal: {
        onUpdate: t?.onUpdate
      },
      [q]: n
    };
  },
  containerProps: ({ id: i, className: e, style: t, full: n, containerProps: r }) => ({
    className: e,
    id: i,
    style: {
      ...n ? {
        position: "absolute",
        bottom: "0",
        left: "0",
        right: "0",
        top: "0"
      } : {},
      ...t
    },
    ...r
  }),
  cesiumProps: Yi,
  cesiumReadonlyProps: Ji,
  otherProps: Ki,
  renderContainer: !0,
  useCommonEvent: !0,
  useRootEvent: !0
}), Zi = [
  "classificationType",
  "debugShowBoundingVolume",
  "debugShowShadowVolume",
  "show"
], Xi = [
  "allowPicking",
  "asynchronous",
  "compressVertices",
  "geometryInstances",
  "interleave",
  "releaseGeometryInstances",
  "vertexCacheOptimize",
  "appearance"
], Qi = ["onReady"], mo = a({
  name: "ClassificationPrimitive",
  async create(i, e) {
    if (!i.primitiveCollection) return;
    const t = new je(e);
    if (e.onReady) {
      const n = () => {
        t.ready && (e.onReady?.(t), i.scene?.postRender.removeEventListener(n));
      };
      i.scene?.postRender.addEventListener(n);
    }
    return i.primitiveCollection.add(t), t;
  },
  destroy(i, e) {
    e.primitiveCollection && !e.primitiveCollection.isDestroyed() && e.primitiveCollection.remove(i), i.isDestroyed() || i.destroy();
  },
  cesiumProps: Zi,
  cesiumReadonlyProps: Xi,
  otherProps: Qi,
  useCommonEvent: !0
}), et = {
  onStop: "onStop",
  onTick: "onTick"
}, it = [
  "canAnimate",
  "clockRange",
  "clockStep",
  "currentTime",
  "multiplier",
  "shouldAnimate",
  "startTime",
  "stopTime"
], ho = a({
  name: "Clock",
  create: (i) => i.cesiumWidget?.clock,
  cesiumProps: it,
  cesiumEventProps: et,
  setCesiumPropsAfterCreate: !0
}), tt = [
  "noiseDetail",
  "noiseOffset",
  "show",
  "debugBillboards",
  "debugEllipsoids"
], fo = a({
  name: "CloudCollection",
  create: (i) => {
    if (!i.primitiveCollection) return;
    const e = new Ne();
    return i.primitiveCollection.add(e), e;
  },
  destroy(i, e) {
    e.primitiveCollection && !e.primitiveCollection.isDestroyed() && e.primitiveCollection.remove(i), i.isDestroyed() || i.destroy();
  },
  provide: (i) => ({
    cloudCollection: i
  }),
  cesiumProps: tt,
  setCesiumPropsAfterCreate: !0
}), nt = [
  "positions",
  "width",
  "cornerType",
  "height",
  "heightReference",
  "extrudedHeight",
  "extrudedHeightReference",
  "show",
  "fill",
  "material",
  "outline",
  "outlineColor",
  "outlineWidth",
  "granularity",
  "shadows",
  "distanceDisplayCondition",
  "zIndex",
  "classificationType"
], ot = {
  onDefinitionChange: "definitionChanged"
}, yo = a({
  name: "CorridorGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new ze(e);
    return e.classificationType && (t.classificationType = e.classificationType), i.entity.corridor = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.corridor = void 0);
  },
  cesiumProps: nt,
  cesiumEventProps: ot
}), rt = [
  "show",
  "position",
  "scale",
  "maximumSize",
  "slice",
  "brightness",
  "color"
], po = a({
  name: "CumulusCloud",
  create: (i, e) => i.cloudCollection?.add(e),
  destroy(i, e) {
    e.cloudCollection && !e.cloudCollection.isDestroyed() && e.cloudCollection.remove(i);
  },
  cesiumProps: rt
}), st = ["clustering", "name", "show", "clock", "isLoading"], at = {
  onChange: "changedEvent",
  onError: "errorEvent",
  onLoading: "loadingEvent"
}, go = a({
  name: "CustomDataSource",
  create(i, e) {
    if (!i.dataSourceCollection) return;
    const t = new xe(e.name);
    return e.clustering && (t.clustering = e.clustering), typeof e.show == "boolean" && (t.show = e.show), typeof e.clock < "u" && (t.clock = e.clock), i.dataSourceCollection.add(t), t;
  },
  destroy(i, e) {
    e.dataSourceCollection && !e.dataSourceCollection.isDestroyed() && e.dataSourceCollection.remove(i);
  },
  provide(i) {
    return {
      entityCollection: i.entities,
      dataSource: i
    };
  },
  cesiumProps: st,
  cesiumEventProps: at,
  useCommonEvent: !0
}), ct = [
  "heightReference",
  "length",
  "topRadius",
  "bottomRadius",
  "show",
  "fill",
  "material",
  "outline",
  "outlineColor",
  "outlineWidth",
  "numberOfVerticalLines",
  "slices",
  "distanceDisplayCondition",
  "shadows"
], lt = {
  onDefinitionChange: "definitionChanged"
}, vo = a({
  name: "CylinderGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new qe(e);
    return i.entity.cylinder = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.cylinder = void 0);
  },
  cesiumProps: ct,
  cesiumEventProps: lt
}), ut = ["clustering", "show"], dt = ["name", "sourceUri", "credit"], mt = {
  onChange: "changedEvent",
  onError: "errorEvent",
  onLoading: "loadingEvent"
}, ht = ["onLoad", "data"], Se = (i, { data: e, onLoad: t, ...n }) => {
  e && i.load(e, n).then((r) => {
    t && t(r);
  });
}, Co = a({
  name: "CzmlDataSource",
  create(i, e) {
    if (!i.dataSourceCollection) return;
    const t = new Ye(e.name);
    return e.clustering && (t.clustering = e.clustering), typeof e.show == "boolean" && (t.show = e.show), i.dataSourceCollection.add(t), e.data && Se(t, e), t;
  },
  update(i, e, t) {
    e.data ? t.show !== e.show && (i.show = typeof e.show == "boolean" ? e.show : !0) : i.show = !1, e.data && (t.data !== e.data || t.sourceUri !== e.sourceUri || t.credit !== e.credit) && Se(i, e);
  },
  destroy(i, e) {
    e.dataSourceCollection && !e.dataSourceCollection.isDestroyed() && e.dataSourceCollection.remove(i);
  },
  provide(i) {
    return {
      dataSource: i
    };
  },
  cesiumProps: ut,
  cesiumReadonlyProps: dt,
  cesiumEventProps: mt,
  otherProps: ht,
  useCommonEvent: !0
}), ft = [
  "semiMajorAxis",
  "semiMinorAxis",
  "height",
  "heightReference",
  "extrudedHeight",
  "show",
  "fill",
  "material",
  "outline",
  "outlineColor",
  "outlineWidth",
  "numberOfVerticalLines",
  "rotation",
  "stRotation",
  "granularity",
  "shadows",
  "distanceDisplayCondition",
  "zIndex",
  "classificationType",
  "extrudedHeightReference"
], yt = {
  onDefinitionChange: "definitionChanged"
}, Po = a({
  name: "EllipseGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new Je(e);
    return i.entity.ellipse = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.ellipse = void 0);
  },
  cesiumProps: ft,
  cesiumEventProps: yt
}), pt = [
  "heightReference",
  "radii",
  "show",
  "fill",
  "innerRadii",
  "material",
  "maximumClock",
  "maximumCone",
  "minimumClock",
  "minimumCone",
  "outline",
  "outlineColor",
  "outlineWidth",
  "subdivisions",
  "stackPartitions",
  "slicePartitions",
  "shadows",
  "distanceDisplayCondition"
], gt = {
  onDefinitionChange: "definitionChanged"
}, Eo = a({
  name: "EllipsoidGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new Ke(e);
    return i.entity.ellipsoid = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.ellipsoid = void 0);
  },
  cesiumProps: pt,
  cesiumEventProps: gt
}), vt = [
  "availability",
  "billboard",
  "box",
  "corridor",
  "cylinder",
  "description",
  "ellipse",
  "ellipsoid",
  "entityCollection",
  "label",
  "model",
  "name",
  "orientation",
  "path",
  "plane",
  "parent",
  "point",
  "polygon",
  "polyline",
  "polylineVolume",
  "position",
  "properties",
  "rectangle",
  "show",
  "tileset",
  "trackingReferenceFrame",
  "viewFrom",
  "wall"
], Ct = ["id"], Pt = {
  onDefinitionChange: "definitionChanged"
}, Et = ["selected", "tracked"], wo = a({
  name: "Entity",
  create(i, e) {
    if (!i.entityCollection) return;
    const t = new Ze(e);
    return i.viewer && e.selected && (i.viewer.selectedEntity = t), i.viewer && e.tracked && (i.viewer.trackedEntity = t), i.entityCollection.add(t), t;
  },
  destroy(i, e) {
    e.entityCollection && e.entityCollection.remove(i);
  },
  update(i, e, t, n) {
    n.viewer && (e.selected !== t.selected && (e.selected ? n.viewer.selectedEntity = i : n.viewer.selectedEntity === i && (n.viewer.selectedEntity = void 0)), e.tracked !== t.tracked && (e.tracked ? n.viewer.trackedEntity = i : n.viewer.trackedEntity === i && (n.viewer.trackedEntity = void 0)));
  },
  provide(i, e, t) {
    return {
      entity: i,
      __$internal: {
        onUpdate: t?.onUpdate
      }
    };
  },
  cesiumProps: vt,
  cesiumReadonlyProps: Ct,
  cesiumEventProps: Pt,
  otherProps: Et,
  useCommonEvent: !0
}), So = ({
  children: i,
  container: e,
  resizeInfoBox: t = !0
}) => {
  const { viewer: n, entity: r } = re(), [c, l] = Te(!1), m = Ge(
    () => e ?? n?.infoBox.frame.contentDocument?.createElement("div"),
    [e, n?.infoBox.frame.contentDocument]
  );
  return Y(() => {
    if (!n || !r) return;
    const u = (y) => {
      l(!!y && y.id === r.id);
    };
    return n.selectedEntityChanged.addEventListener(u), () => {
      n.selectedEntityChanged.removeEventListener(u);
    };
  }, [r, n]), Y(() => {
    if (e || !m || !n) return;
    const u = n.infoBox?.frame, y = u?.contentDocument?.querySelector(".cesium-infoBox-description");
    if (!u || !y) return;
    let P;
    if (c) {
      if (t) {
        const w = y.getBoundingClientRect().height;
        u.style.height = w + "px", P = window.setInterval(() => {
          const S = n.infoBox.container.querySelector(
            ".cesium-infoBox.cesium-infoBox-visible"
          );
          S && (clearInterval(P), P = void 0, y.appendChild(m), S.classList.remove("cesium-infoBox-bodyless"), u.style.height = y.getBoundingClientRect().height + "px");
        }, 10);
      }
    } else m.parentElement === y && y.removeChild(m);
    return P ? () => clearTimeout(P) : void 0;
  }, [m, e, t, c, n]), m ? /* @__PURE__ */ $.jsx($.Fragment, { children: Si(/* @__PURE__ */ $.jsx($.Fragment, { children: !e || c ? i : null }), m) }) : null;
}, wt = [
  "density",
  "enabled",
  "heightFalloff",
  "heightScalar",
  "maxHeight",
  "minimumBrightness",
  "renderable",
  "screenSpaceErrorFactor",
  "visualDensityScalar"
], bo = a({
  name: "Fog",
  create(i) {
    if (!i.scene) return;
    const e = new ye();
    return i.scene.fog = e, e;
  },
  destroy(i, e) {
    e.scene && !e.scene.isDestroyed() && (e.scene.fog = new ye());
  },
  cesiumProps: wt,
  setCesiumPropsAfterCreate: !0
}), St = ["clustering", "name", "show"], bt = [
  "clampToGround",
  "sourceUri",
  "credit",
  "markerSize",
  "markerSymbol",
  "markerColor",
  "stroke",
  "strokeWidth",
  "fill",
  "describe"
], Dt = {
  onChange: "changedEvent",
  onError: "errorEvent",
  onLoading: "loadingEvent"
}, Rt = ["onLoad", "data"], be = (i, { data: e, onLoad: t, ...n }) => {
  e && i.load(e, n).then((r) => {
    t && t(r);
  });
}, Do = a({
  name: "GeoJsonDataSource",
  create(i, e) {
    if (!i.dataSourceCollection) return;
    const t = new Xe(e.name);
    return e.clustering && (t.clustering = e.clustering), typeof e.show == "boolean" && (t.show = e.show), i.dataSourceCollection.add(t), e.data && be(t, e), t;
  },
  update(i, e, t) {
    e.data ? t.show !== e.show && (i.show = typeof e.show == "boolean" ? e.show : !0) : i.show = !1, e.data && (t.data !== e.data || t.clampToGround !== e.clampToGround || t.sourceUri !== e.sourceUri || t.credit !== e.credit || t.markerSize !== e.markerSize || t.markerSymbol !== e.markerSymbol || t.markerColor !== e.markerColor || t.stroke !== e.stroke || t.strokeWidth !== e.strokeWidth || t.fill !== e.fill) && be(i, e);
  },
  destroy(i, e) {
    e.dataSourceCollection && !e.dataSourceCollection.isDestroyed() && e.dataSourceCollection.remove(i);
  },
  provide(i) {
    return {
      dataSource: i
    };
  },
  cesiumProps: St,
  cesiumReadonlyProps: bt,
  cesiumEventProps: Dt,
  otherProps: Rt,
  useCommonEvent: !0
}), Tt = {
  onImageryLayersUpdate: "imageryLayersUpdatedEvent",
  onTerrainProviderChange: "terrainProviderChanged",
  onTileLoadProgress: "tileLoadProgressEvent"
}, Mt = [
  "atmosphereBrightnessShift",
  "atmosphereHueShift",
  "atmosphereSaturationShift",
  "backFaceCulling",
  "baseColor",
  "clippingPlanes",
  "depthTestAgainstTerrain",
  "enableLighting",
  "lightingFadeInDistance",
  "lightingFadeOutDistance",
  "material",
  "maximumScreenSpaceError",
  "nightFadeInDistance",
  "nightFadeOutDistance",
  "oceanNormalMapUrl",
  "shadows",
  "show",
  "showGroundAtmosphere",
  "showWaterEffect",
  "tileCacheSize",
  "loadingDescendantLimit",
  "preloadAncestors",
  "preloadSiblings",
  "fillHighlightColor",
  "dynamicAtmosphereLighting",
  "dynamicAtmosphereLightingFromSun",
  "showSkirts",
  "cartographicLimitRectangle",
  "translucency",
  "undergroundColor",
  "undergroundColorAlphaByDistance",
  "lambertDiffuseMultiplier",
  "atmosphereLightIntensity",
  "atmosphereRayleighCoefficient",
  "atmosphereMieCoefficient",
  "atmosphereRayleighScaleHeight",
  "atmosphereMieScaleHeight",
  "atmosphereMieAnisotropy",
  "vertexShadowDarkness",
  "clippingPolygons"
], kt = ["terrainProvider"], Ro = a({
  name: "Globe",
  create: (i) => i.scene?.globe,
  update: async (i, e) => {
    const t = e.terrainProvider;
    let n;
    I(t) ? n = await t : n = t, i.terrainProvider = n;
  },
  cesiumProps: Mt,
  cesiumEventProps: Tt,
  otherProps: kt,
  setCesiumPropsAfterCreate: !0
}), To = a({
  name: "Google2DImageryProvider",
  async create(i, e) {
    const { onReady: t, assetId: n, accessToken: r, key: c, children: l, ...m } = e;
    let u;
    if (n !== void 0) {
      const y = { ...m, assetId: n };
      r !== void 0 && (y.accessToken = r), u = await pe.fromIonAssetId(y);
    } else {
      const y = { ...m };
      c !== void 0 && (y.key = c), u = await pe.fromUrl(y);
    }
    return t && t(u), u;
  }
}), $t = [
  "appearance",
  "classificationType",
  "debugShowBoundingVolume",
  "debugShowShadowVolume",
  "show"
], Lt = [
  "allowPicking",
  "asynchronous",
  "geometryInstances",
  "interleave",
  "releaseGeometryInstances"
], _t = ["onReady"], Mo = a({
  name: "GroundPolylinePrimitive",
  create(i, e) {
    if (!i.primitiveCollection) return;
    const t = new Qe(e);
    if (e.onReady) {
      const n = () => {
        t.ready && (e.onReady?.(t), i.scene?.postRender.removeEventListener(n));
      };
      i.scene?.postRender.addEventListener(n);
    }
    return i.primitiveCollection.add(t), t;
  },
  destroy(i, e) {
    e.primitiveCollection && !e.primitiveCollection.isDestroyed() && e.primitiveCollection.remove(i), i.isDestroyed() || i.destroy();
  },
  cesiumProps: $t,
  cesiumReadonlyProps: Lt,
  otherProps: _t,
  useCommonEvent: !0
}), Ot = [
  "appearance",
  "classificationType",
  "debugShowBoundingVolume",
  "debugShowShadowVolume",
  "show"
], At = [
  "allowPicking",
  "asynchronous",
  "compressVertices",
  "geometryInstances",
  "interleave",
  "releaseGeometryInstances",
  "vertexCacheOptimize"
], Bt = ["onReady"], ko = a({
  name: "GroundPrimitive",
  create(i, e) {
    if (!i.primitiveCollection) return;
    const t = new ei(e);
    if (e.onReady) {
      const n = () => {
        t.ready && (e.onReady?.(t), i.scene?.postRender.removeEventListener(n));
      };
      i.scene?.postRender.addEventListener(n);
    }
    return i.primitiveCollection.add(t), t;
  },
  destroy(i, e) {
    e.primitiveCollection && !e.primitiveCollection.isDestroyed() && e.primitiveCollection.remove(i), i.isDestroyed() || i.destroy();
  },
  cesiumProps: Ot,
  cesiumReadonlyProps: At,
  otherProps: Bt,
  useCommonEvent: !0
}), Gt = ["show", "destroyPrimitives", "primitiveAdded", "primitiveRemoved"], $o = a({
  name: "GroundPrimitiveCollection",
  create: (i) => i.scene?.groundPrimitives,
  provide: (i) => ({
    primitiveCollection: i
  }),
  cesiumProps: Gt,
  setCesiumPropsAfterCreate: !0
}), It = [
  "alpha",
  "brightness",
  "contrast",
  "hue",
  "saturation",
  "gamma",
  "splitDirection",
  "minificationFilter",
  "magnificationFilter",
  "cutoutRectangle",
  "show",
  "nightAlpha",
  "dayAlpha",
  "colorToAlpha",
  "colorToAlphaThreshold",
  "index"
], Ft = [
  "rectangle",
  "maximumAnisotropy",
  "minimumTerrainLevel",
  "maximumTerrainLevel",
  "imageryProvider"
], Lo = a({
  name: "ImageryLayer",
  async create(i, e) {
    if (!i.imageryLayerCollection) return;
    const t = I(e.imageryProvider) ? e.imageryProvider : new Promise((l) => queueMicrotask(() => l(e.imageryProvider))), n = i.__$internal?.imageryLayerWaitingList?.slice();
    i.__$internal?.imageryLayerWaitingList && i.__$internal.imageryLayerWaitingList.push(t), n && await Promise.all(n.filter((l) => I(l)));
    const r = await t;
    if (i.__$internal?.imageryLayerWaitingList && (i.__$internal.imageryLayerWaitingList = i.__$internal.imageryLayerWaitingList.filter((l) => l !== t)), !r) return;
    const c = new ii(r, e);
    return i.imageryLayerCollection.add(c, e.index), c;
  },
  destroy(i, e) {
    e.imageryLayerCollection && e.imageryLayerCollection.remove(i);
  },
  cesiumProps: It,
  cesiumReadonlyProps: Ft
}), Vt = {
  onLayerAdd: "layerAdded",
  onLayerMove: "layerMoved",
  onLayerRemove: "layerRemoved",
  onLayerShowOrHide: "layerShownOrHidden"
}, _o = a({
  name: "ImageryLayerCollection",
  create: (i) => i.globe?.imageryLayers,
  cesiumEventProps: Vt
}), Ut = ["clustering", "name", "show"], Wt = [
  "canvas",
  "camera",
  "ellipsoid",
  "clampToGround",
  "sourceUri",
  "credit",
  "screenOverlayContainer"
], Ht = {
  onChange: "changedEvent",
  onError: "errorEvent",
  onLoading: "loadingEvent",
  onRefresh: "refreshEvent",
  onUnsupportedNode: "unsupportedNodeEvent"
}, jt = ["onLoad", "data"], De = (i, { data: e, onLoad: t, ...n }) => {
  e && i.load(e, n).then((r) => {
    t && t(r);
  });
}, Oo = a({
  name: "KmlDataSource",
  create(i, e) {
    if (!i.scene || !i.dataSourceCollection || !i.scene) return;
    const t = new ti({
      camera: e.camera || i.scene.camera,
      canvas: e.canvas || i.scene.canvas,
      ellipsoid: e.ellipsoid,
      credit: e.credit
    });
    return e.clustering && (t.clustering = e.clustering), typeof e.show == "boolean" && (t.show = e.show), typeof e.name < "u" && (t.name = e.name), i.dataSourceCollection.add(t), e.data && De(t, e), t;
  },
  update(i, e, t) {
    e.data ? t.show !== e.show && (i.show = typeof e.show == "boolean" ? e.show : !0) : i.show = !1, e.data && (t.data !== e.data || t.clampToGround !== e.clampToGround || t.ellipsoid !== e.ellipsoid || t.sourceUri !== e.sourceUri || t.credit !== e.credit) && De(i, e);
  },
  destroy(i, e) {
    e.dataSourceCollection && !e.dataSourceCollection.isDestroyed() && e.dataSourceCollection.remove(i);
  },
  provide(i) {
    return {
      dataSource: i
    };
  },
  cesiumProps: Ut,
  cesiumReadonlyProps: Wt,
  cesiumEventProps: Ht,
  otherProps: jt,
  useCommonEvent: !0
}), Nt = [
  "backgroundColor",
  "backgroundPadding",
  "disableDepthTestDistance",
  "distanceDisplayCondition",
  "eyeOffset",
  "fillColor",
  "font",
  "heightReference",
  "horizontalOrigin",
  "id",
  "outlineColor",
  "outlineWidth",
  "pixelOffset",
  "pixelOffsetScaleByDistance",
  "position",
  "scale",
  "scaleByDistance",
  "show",
  "showBackground",
  "style",
  "text",
  "translucencyByDistance",
  "verticalOrigin"
], Ao = a({
  name: "Label",
  create: (i, e) => i.labelCollection?.add(e),
  destroy(i, e) {
    e.labelCollection && !e.labelCollection.isDestroyed() && e.labelCollection.remove(i);
  },
  cesiumProps: Nt,
  useCommonEvent: !0
}), zt = ["blendOption", "debugShowBoundingVolume", "modelMatrix", "show"], Bo = a({
  name: "LabelCollection",
  create(i, e) {
    if (!i.scene || !i.primitiveCollection) return;
    const t = new ni({
      scene: i.scene,
      modelMatrix: e.modelMatrix,
      blendOption: e.blendOption,
      debugShowBoundingVolume: e.debugShowBoundingVolume
    });
    return i.primitiveCollection.add(t), t;
  },
  destroy(i, e) {
    e.primitiveCollection && !e.primitiveCollection.isDestroyed() && e.primitiveCollection.remove(i), i.isDestroyed() || i.destroy();
  },
  provide(i) {
    return {
      labelCollection: i
    };
  },
  cesiumProps: zt
}), xt = [
  "text",
  "font",
  "style",
  "fillColor",
  "outlineColor",
  "outlineWidth",
  "show",
  "showBackground",
  "backgroundColor",
  "backgroundPadding",
  "scale",
  "horizontalOrigin",
  "verticalOrigin",
  "eyeOffset",
  "pixelOffset",
  "translucencyByDistance",
  "pixelOffsetScaleByDistance",
  "scaleByDistance",
  "heightReference",
  "distanceDisplayCondition",
  "disableDepthTestDistance"
], qt = {
  onDefinitionChange: "definitionChanged"
}, Go = a({
  name: "LabelGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new oi(e);
    return i.entity.label = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.label = void 0);
  },
  cesiumProps: xt,
  cesiumEventProps: qt
}), Yt = [
  "backFaceCulling",
  "clampAnimations",
  "clippingPlanes",
  "clippingPolygons",
  "color",
  "colorBlendAmount",
  "colorBlendMode",
  "customShader",
  "debugShowBoundingVolume",
  "debugWireframe",
  "distanceDisplayCondition",
  "enableVerticalExaggeration",
  "featureIdLabel",
  "heightReference",
  "id",
  "imageBasedLighting",
  "instanceFeatureIdLabel",
  "lightColor",
  "maximumScale",
  "minimumPixelSize",
  "modelMatrix",
  "outlineColor",
  "scale",
  "shadows",
  "show",
  "showCreditsOnScreen",
  "showOutline",
  "silhouetteColor",
  "silhouetteSize",
  "splitDirection",
  "style",
  "pointCloudShading"
], Jt = [
  "allowPicking",
  "asynchronous",
  "basePath",
  "credit",
  "enableDebugWireframe",
  "gltf",
  "incrementallyLoadTextures",
  "scene",
  "releaseGltfJson",
  "cull",
  "opaquePass",
  "upAxis",
  "forwardAxis",
  "content",
  "enableShowOutline",
  "projectTo2D",
  "classificationType",
  "gltfCallback",
  "enablePick",
  "environmentMapOptions"
], Kt = ["onReady", "onError", "url"], Io = a({
  name: "Model",
  async create(i, { scene: e, url: t, colorBlendMode: n, ...r }) {
    if (!i.scene || !i.primitiveCollection || !t) return;
    const c = t;
    let l;
    I(c) ? l = await c : l = c;
    let m;
    try {
      m = await ri.fromGltfAsync({
        ...r,
        url: l,
        colorBlendMode: n,
        scene: e || i.scene
      });
    } catch (u) {
      r.onError?.(u);
      return;
    }
    return i.primitiveCollection.add(m), m;
  },
  destroy(i, e) {
    e.primitiveCollection && !e.primitiveCollection.isDestroyed() && e.primitiveCollection.remove(i), i.isDestroyed() || i.destroy();
  },
  cesiumEventProps: {
    onReady: "readyEvent"
  },
  cesiumProps: Yt,
  cesiumReadonlyProps: Jt,
  otherProps: Kt,
  useCommonEvent: !0
}), Zt = [
  "uri",
  "show",
  "scale",
  "minimumPixelSize",
  "maximumScale",
  "incrementallyLoadTextures",
  "runAnimations",
  "clampAnimations",
  "nodeTransformations",
  "shadows",
  "heightReference",
  "distanceDisplayCondition",
  "silhouetteColor",
  "silhouetteSize",
  "color",
  "colorBlendMode",
  "colorBlendAmount",
  "clippingPlanes",
  "imageBasedLightingFactor",
  "lightColor",
  "articulations",
  "customShader",
  "enableVerticalExaggeration",
  "environmentMapOptions"
], Xt = {
  onDefinitionChange: "definitionChanged"
}, Fo = a({
  name: "ModelGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new si(e);
    return i.entity.model = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.model = void 0);
  },
  cesiumProps: Zt,
  cesiumEventProps: Xt
}), Qt = ["onlySunLighting", "show", "textureUrl"], en = ["ellipsoid"], Vo = a({
  name: "Moon",
  create(i, e) {
    if (!i.scene) return;
    const t = new ge(e);
    return i.scene.moon = t, t;
  },
  destroy(i, e) {
    e.scene && !e.scene.isDestroyed() && (e.scene.moon = new ge());
  },
  cesiumProps: Qt,
  cesiumReadonlyProps: en
}), tn = [
  "show",
  "emitter",
  "modelMatrix",
  "emitterModelMatrix",
  "emissionRate",
  "bursts",
  "loop",
  "startScale",
  "endScale",
  "startColor",
  "endColor",
  "image",
  "imageSize",
  "minimumImageSize",
  "maximumImageSize",
  "speed",
  "minimumSpeed",
  "maximumSpeed",
  "lifetime",
  "particleLife",
  "minimumParticleLife",
  "maximumParticleLife",
  "mass",
  "minimumMass",
  "maximumMass",
  "sizeInMeters"
], nn = {
  onComplete: "complete",
  onUpdate: "updateCallback"
}, Uo = a({
  name: "ParticleSystem",
  create(i, e) {
    if (!i.primitiveCollection) return;
    const t = new ai({
      ...e,
      updateCallback: e.onUpdate
    });
    return i.primitiveCollection.add(t), t;
  },
  update(i, e, t) {
    e.onUpdate !== t.onUpdate && (i.updateCallback = e.onUpdate);
  },
  destroy(i, e) {
    e.primitiveCollection && !e.primitiveCollection.isDestroyed() && e.primitiveCollection.remove(i);
  },
  cesiumProps: tn,
  cesiumEventProps: nn
}), on = [
  "leadTime",
  "trailTime",
  "show",
  "width",
  "material",
  "resolution",
  "distanceDisplayCondition"
], rn = {
  onDefinitionChange: "definitionChanged"
}, Wo = a({
  name: "PathGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new ci(e);
    return i.entity.path = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.path = void 0);
  },
  cesiumProps: on,
  cesiumEventProps: rn
}), sn = [
  "plane",
  "dimensions",
  "show",
  "fill",
  "material",
  "outline",
  "outlineColor",
  "outlineWidth",
  "shadows",
  "distanceDisplayCondition"
], an = {
  onDefinitionChange: "definitionChanged"
}, Ho = a({
  name: "PlaneGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new li(e);
    return i.entity.plane = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.plane = void 0);
  },
  cesiumProps: sn,
  cesiumEventProps: an
}), cn = [
  "color",
  "pixelSize",
  "outlineColor",
  "outlineWidth",
  "show",
  "scaleByDistance",
  "translucencyByDistance",
  "heightReference",
  "distanceDisplayCondition",
  "disableDepthTestDistance",
  "splitDirection"
], ln = {
  onDefinitionChange: "definitionChanged"
}, jo = a({
  name: "PointGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new ui(e);
    return i.entity.point = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.point = void 0);
  },
  cesiumProps: cn,
  cesiumEventProps: ln
}), un = [
  "color",
  "disableDepthTestDistance",
  "distanceDisplayCondition",
  "id",
  "outlineColor",
  "outlineWidth",
  "pixelSize",
  "position",
  "scaleByDistance",
  "show",
  "splitDirection",
  "translucencyByDistance"
], No = a({
  name: "PointPrimitive",
  create: (i, e) => i.pointPrimitiveCollection?.add(e),
  destroy(i, e) {
    e.pointPrimitiveCollection && !e.pointPrimitiveCollection.isDestroyed() && e.pointPrimitiveCollection.remove(i);
  },
  cesiumProps: un,
  useCommonEvent: !0
}), dn = ["blendOption", "debugShowBoundingVolume", "modelMatrix", "show"], zo = a({
  name: "PointPrimitveCollection",
  create(i, e) {
    if (!i.primitiveCollection) return;
    const t = new di(e);
    return i.primitiveCollection.add(t), t;
  },
  destroy(i, e) {
    e.primitiveCollection && !e.primitiveCollection.isDestroyed() && e.primitiveCollection.remove(i), i.isDestroyed() || i.destroy();
  },
  provide(i) {
    return {
      pointPrimitiveCollection: i
    };
  },
  cesiumProps: dn
}), mn = [
  "arcType",
  "hierarchy",
  "height",
  "heightReference",
  "extrudedHeight",
  "extrudedHeightReference",
  "show",
  "fill",
  "material",
  "outline",
  "outlineColor",
  "outlineWidth",
  "stRotation",
  "granularity",
  "perPositionHeight",
  "closeTop",
  "closeBottom",
  "shadows",
  "distanceDisplayCondition",
  "zIndex",
  "classificationType",
  "textureCoordinates"
], hn = {
  onDefinitionChange: "definitionChanged"
}, xo = a({
  name: "PolygonGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new mi(e);
    return i.entity.polygon = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.polygon = void 0);
  },
  cesiumProps: mn,
  cesiumEventProps: hn
}), fn = [
  "distanceDisplayCondition",
  "id",
  "loop",
  "material",
  "positions",
  "show",
  "width"
], qo = a({
  name: "Polyline",
  create: (i, e) => i.polylineCollection?.add(e),
  destroy(i, e) {
    e.polylineCollection && !e.polylineCollection.isDestroyed() && e.polylineCollection.remove(i);
  },
  cesiumProps: fn,
  useCommonEvent: !0
}), yn = ["debugShowBoundingVolume", "length", "modelMatrix", "show"], Yo = a(
  {
    name: "PolylineCollection",
    create(i, e) {
      if (!i.primitiveCollection) return;
      const t = new hi({
        modelMatrix: e.modelMatrix,
        debugShowBoundingVolume: e.debugShowBoundingVolume,
        length: e.length,
        scene: i.scene
      });
      return i.primitiveCollection.add(t), t;
    },
    destroy(i, e) {
      e.primitiveCollection && !e.primitiveCollection.isDestroyed() && e.primitiveCollection.remove(i), i.isDestroyed() || i.destroy();
    },
    provide(i) {
      return {
        polylineCollection: i
      };
    },
    cesiumProps: yn
  }
), pn = [
  "arcType",
  "classificationType",
  "positions",
  "clampToGround",
  "width",
  "show",
  "material",
  "depthFailMaterial",
  "granularity",
  "shadows",
  "distanceDisplayCondition",
  "zIndex"
], gn = {
  onDefinitionChange: "definitionChanged"
}, Jo = a({
  name: "PolylineGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new fi(e);
    return i.entity.polyline = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.polyline = void 0);
  },
  cesiumProps: pn,
  cesiumEventProps: gn
}), vn = [
  "positions",
  "shape",
  "cornerType",
  "show",
  "fill",
  "material",
  "outline",
  "outlineColor",
  "outlineWidth",
  "granularity",
  "shadows",
  "distanceDisplayCondition"
], Cn = {
  onDefinitionChange: "definitionChanged"
}, Ko = a({
  name: "PolylineVolumeGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new yi(e);
    return i.entity.polylineVolume = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.polylineVolume = void 0);
  },
  cesiumProps: vn,
  cesiumEventProps: Cn
}), Pn = ["enabled", "selected"], En = [
  "clearColor",
  "forcePowerOfTwo",
  "fragmentShader",
  "name",
  "pixelDatatype",
  "pixelFormat",
  "sampleMode",
  "scissorRectangle",
  "textureScale",
  "uniforms"
], Zo = a({
  name: "PostProcessStage",
  create(i, e) {
    if (!i.scene) return;
    const t = new pi(e);
    return typeof e.enabled == "boolean" && (t.enabled = e.enabled), e.selected && (t.selected = e.selected), i.scene.postProcessStages.add(t), t;
  },
  destroy(i, e) {
    e.scene && !e.scene.isDestroyed() && e.scene.postProcessStages.remove(i), i.isDestroyed() || i.destroy();
  },
  cesiumProps: Pn,
  cesiumReadonlyProps: En
}), Xo = T({
  name: "BlackAndWhiteStage",
  props: ["gradations"],
  create: () => F.createBlackAndWhiteStage()
}), Qo = T({
  name: "BrightnessStage",
  props: ["brightness"],
  create: () => F.createBrightnessStage()
}), er = T({
  name: "LensFlareStage",
  props: [
    "dirtTexture",
    "starTexture",
    "intensity",
    "distortion",
    "ghostDispersal",
    "haloWidth",
    "earthRadius"
  ],
  create: () => F.createLensFlareStage()
}), ir = T({
  name: "Fxaa",
  create: (i, e) => e.fxaa,
  props: []
}), tr = T({
  name: "NightVisionStage",
  props: [],
  create: () => F.createNightVisionStage()
}), wn = ["enabled", "selected"], Sn = ["inputPreviousStageTexture", "name", "stages", "uniforms"], nr = a({
  name: "PostProcessStageComposite",
  create(i, e) {
    if (!i.scene) return;
    const t = new gi(e);
    return typeof e.enabled == "boolean" && (t.enabled = e.enabled), e.selected && (t.selected = e.selected), i.scene.postProcessStages.add(t), t;
  },
  destroy(i, e) {
    e.scene && !e.scene.isDestroyed() && e.scene.postProcessStages.remove(i), i.isDestroyed() || i.destroy();
  },
  cesiumProps: wn,
  cesiumReadonlyProps: Sn
}), or = T({
  name: "AmbientOcclusion",
  create: (i, e) => e.ambientOcclusion,
  props: [
    "ambientOcclusionOnly",
    "bias",
    "delta",
    "frustumLength",
    "intensity",
    "lengthCap",
    "sigma",
    "stepSize"
  ],
  noMount: !0
}), rr = T({
  name: "Bloom",
  create: (i, e) => e.bloom,
  props: ["brightness", "contrast", "delta", "glowOnly", "sigma", "stepSize"],
  noMount: !0
}), sr = T({
  name: "BlurStage",
  props: ["delta", "sigma", "stepSize"],
  create: () => F.createBlurStage()
}), ar = T({
  name: "DepthOfFieldStage",
  props: ["delta", "focalDistance", "sigma", "stepSize"],
  create: () => F.createDepthOfFieldStage()
}), cr = T({
  name: "EdgeDetectionStage",
  props: ["color", "length"],
  create: () => F.createEdgeDetectionStage()
}), lr = T({
  name: "SilhouetteStage",
  props: ["color", "length"],
  create: () => F.createSilhouetteStage()
}), bn = [
  "appearance",
  "cull",
  "debugShowBoundingVolume",
  "depthFailAppearance",
  "modelMatrix",
  "shadows",
  "show"
], Dn = [
  "allowPicking",
  "asynchronous",
  "compressVertices",
  "geometryInstances",
  "interleave",
  "releaseGeometryInstances",
  "vertexCacheOptimize"
], Rn = ["onReady"], ur = a({
  name: "Primitive",
  create(i, e) {
    if (!i.primitiveCollection) return;
    const t = new vi(e);
    if (e.onReady) {
      const n = () => {
        t.ready && (e.onReady?.(t), i.scene?.postRender.removeEventListener(n));
      };
      i.scene?.postRender.addEventListener(n);
    }
    return i.primitiveCollection.add(t), t;
  },
  destroy(i, e) {
    e.primitiveCollection && !e.primitiveCollection.isDestroyed() && e.primitiveCollection.remove(i), i.isDestroyed() || i.destroy();
  },
  cesiumProps: bn,
  cesiumReadonlyProps: Dn,
  otherProps: Rn,
  useCommonEvent: !0
}), Tn = [
  "classificationType",
  "coordinates",
  "height",
  "heightReference",
  "extrudedHeight",
  "extrudedHeightReference",
  "show",
  "fill",
  "material",
  "outline",
  "outlineColor",
  "outlineWidth",
  "rotation",
  "stRotation",
  "granularity",
  "shadows",
  "distanceDisplayCondition",
  "zIndex"
], Mn = {
  onDefinitionChange: "definitionChanged"
}, dr = a({
  name: "RectangleGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new Ci(e);
    return i.entity.rectangle = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.rectangle = void 0);
  },
  cesiumProps: Tn,
  cesiumEventProps: Mn
}), kn = [
  "backgroundColor",
  "completeMorphOnUserInput",
  // "debugCommandFilter", // Ignored in type check
  "debugShowCommands",
  "debugShowDepthFrustum",
  "debugShowFramesPerSecond",
  "debugShowFrustumPlanes",
  "debugShowFrustums",
  "eyeSeparation",
  "farToNearRatio",
  "focalLength",
  "fog",
  "gamma",
  "globe",
  "highDynamicRange",
  "invertClassification",
  "invertClassificationColor",
  "light",
  "logarithmicDepthBuffer",
  "logarithmicDepthFarToNearRatio",
  "maximumRenderTimeChange",
  "minimumDisableDepthTestDistance",
  // "mode", // enable morph with animation
  "moon",
  "morphTime",
  "nearToFarDistance2D",
  "pickTranslucentDepth",
  "requestRenderMode",
  "rethrowRenderErrors",
  "shadowMap",
  "skyAtmosphere",
  "skyBox",
  "specularEnvironmentMaps",
  "sphericalHarmonicCoefficients",
  "sun",
  "sunBloom",
  "terrainProvider",
  "useDepthPicking",
  "useWebVR",
  // "postProcessStages", // Ignored in type check - readonly property
  "msaaSamples",
  "splitPosition",
  "verticalExaggeration",
  "verticalExaggerationRelativeHeight",
  "atmosphere"
], $n = {
  onMorphComplete: "morphComplete",
  onMorphStart: "morphStart",
  onPostRender: "postRender",
  onPreRender: "preRender",
  onPreUpdate: "preUpdate",
  onPostUpdate: "postUpdate",
  onRenderError: "renderError",
  onTerrainProviderChange: "terrainProviderChanged"
}, Ln = ["mode", "morphDuration", "enableEdgeVisibility"], Re = (i, e, t) => {
  switch (e) {
    case le.SCENE2D:
      i.morphTo2D(t);
      break;
    case le.COLUMBUS_VIEW:
      i.morphToColumbusView(t);
      break;
    case le.SCENE3D:
      i.morphTo3D(t);
      break;
  }
}, mr = a({
  name: "Scene",
  create(i, e) {
    return i.scene && (e.mode && Re(i.scene, e.mode, e.morphDuration), e.enableEdgeVisibility !== void 0 && (i.scene._enableEdgeVisibility = e.enableEdgeVisibility)), i.scene;
  },
  update(i, e, t) {
    e.mode !== t.mode && e.mode && Re(i, e.mode, e.morphDuration), e.enableEdgeVisibility !== t.enableEdgeVisibility && e.enableEdgeVisibility !== void 0 && (i._enableEdgeVisibility = e.enableEdgeVisibility);
  },
  cesiumProps: kn,
  cesiumEventProps: $n,
  otherProps: Ln,
  setCesiumPropsAfterCreate: !0
}), _n = [
  "bounceAnimationTime",
  "enableCollisionDetection",
  "enableInputs",
  "enableLook",
  "enableRotate",
  "enableTilt",
  "enableTranslate",
  "enableZoom",
  "inertiaSpin",
  "inertiaTranslate",
  "inertiaZoom",
  "lookEventTypes",
  "maximumMovementRatio",
  "maximumZoomDistance",
  "minimumCollisionTerrainHeight",
  "minimumPickingTerrainHeight",
  "minimumTrackBallHeight",
  "minimumZoomDistance",
  "rotateEventTypes",
  "tiltEventTypes",
  "translateEventTypes",
  "zoomEventTypes",
  "minimumPickingTerrainDistanceWithInertia",
  "maximumTiltAngle",
  "zoomFactor"
], hr = a({
  name: "ScreenSpaceCameraController",
  create: (i) => i.scene?.screenSpaceCameraController,
  cesiumProps: _n,
  setCesiumPropsAfterCreate: !0
}), fr = ({ action: i, modifier: e, type: t }) => {
  const n = re();
  return Y(() => {
    if (!(!n.screenSpaceEventHandler || n.screenSpaceEventHandler.isDestroyed())) {
      if (i)
        return n.screenSpaceEventHandler.setInputAction(i, t, e), () => {
          !n.screenSpaceEventHandler || n.screenSpaceEventHandler.isDestroyed() || n.screenSpaceEventHandler.removeInputAction(t, e);
        };
      n.screenSpaceEventHandler.removeInputAction(t, e);
    }
  }, [i, n.screenSpaceEventHandler, e, t]), null;
}, yr = a({
  name: "ScreenSpaceEventHandler",
  create(i, e) {
    return e.useDefault ? i.cesiumWidget?.screenSpaceEventHandler : i.scene ? new ke(i.scene.canvas) : void 0;
  },
  destroy(i) {
    i.isDestroyed() || i.destroy();
  },
  provide(i) {
    return {
      screenSpaceEventHandler: i
    };
  }
}), On = [
  "darkness",
  "fadingEnabled",
  "maximumDistance",
  "enabled",
  "normalOffset",
  "softShadows",
  "size"
], pr = a({
  name: "ShadowMap",
  create: (i) => i.scene?.shadowMap,
  cesiumProps: On
}), An = [
  "brightnessShift",
  "hueShift",
  "saturationShift",
  "show",
  "perFragmentAtmosphere",
  "atmosphereLightIntensity",
  "atmosphereRayleighCoefficient",
  "atmosphereMieCoefficient",
  "atmosphereRayleighScaleHeight",
  "atmosphereMieScaleHeight",
  "atmosphereMieAnisotropy"
], gr = a({
  name: "SkyAtmosphere",
  create: (i) => i.scene?.skyAtmosphere,
  cesiumProps: An,
  setCesiumPropsAfterCreate: !0
}), Bn = ["sources", "show"], vr = a({
  name: "SkyBox",
  create: (i) => i.scene?.skyBox,
  cesiumProps: Bn,
  setCesiumPropsAfterCreate: !0
}), Gn = ["glowFactor", "show"], Cr = a({
  name: "Sun",
  create(i) {
    if (!i.scene) return;
    const e = new ve();
    return i.scene.sun = e, e;
  },
  destroy(i, e) {
    e.scene && !e.scene.isDestroyed() && (e.scene.sun = new ve());
  },
  cesiumProps: Gn,
  setCesiumPropsAfterCreate: !0
}), In = [
  "clippingPlanes",
  "maximumMemoryUsage",
  "modelMatrix",
  "shadows",
  "show",
  "style",
  "intervals"
], Fn = ["clock", "shading"], Vn = ["onReady"], Un = {
  onFrameChange: "frameChanged"
}, Pr = a({
  name: "TimeDynamicPointCloud",
  create(i, e) {
    if (!i.cesiumWidget || !i.primitiveCollection || !i.cesiumWidget?.clock)
      return;
    const t = new Pi({
      ...e,
      clock: e.clock ?? i.cesiumWidget.clock
    });
    if (e.onReady) {
      const n = () => {
        e.onReady?.(t), t.frameChanged.removeEventListener(n);
      };
      t.frameChanged.addEventListener(n);
    }
    return i.primitiveCollection.add(t), t;
  },
  destroy(i, e) {
    e.primitiveCollection && !e.primitiveCollection.isDestroyed() && e.primitiveCollection.remove(i), i.isDestroyed() || i.destroy();
  },
  cesiumProps: In,
  cesiumReadonlyProps: Fn,
  cesiumEventProps: Un,
  otherProps: Vn,
  useCommonEvent: !0
}), Wn = [
  "terrainShadows",
  "clockTrackedDataSource",
  "targetFrameRate",
  "useDefaultRenderLoop",
  "resolutionScale",
  "allowDataSourcesToSuspendAnimation",
  "trackedEntity",
  "selectedEntity",
  "shadows",
  "useBrowserRecommendedResolution",
  "creditDisplay"
], Hn = [
  "baseLayer",
  "animation",
  "baseLayerPicker",
  "fullscreenButton",
  "vrButton",
  "geocoder",
  "homeButton",
  "infoBox",
  "sceneModePicker",
  "selectionIndicator",
  "timeline",
  "navigationHelpButton",
  "navigationInstructionsInitiallyVisible",
  "scene3DOnly",
  "shouldAnimate",
  "clockViewModel",
  "selectedImageryProviderViewModel",
  "imageryProviderViewModels",
  "selectedTerrainProviderViewModel",
  "terrainProviderViewModels",
  "ellipsoid",
  "skyBox",
  "skyAtmosphere",
  "fullscreenElement",
  "showRenderLoopErrors",
  "automaticallyTrackDataSourceClocks",
  "contextOptions",
  "sceneMode",
  "mapProjection",
  "globe",
  "orderIndependentTranslucency",
  "creditContainer",
  "creditViewport",
  "dataSources",
  "mapMode2D",
  "projectionPicker",
  "blurActiveElementOnCanvasFocus",
  "requestRenderMode",
  "maximumRenderTimeChange",
  "depthPlaneEllipsoidOffset",
  "msaaSamples",
  "terrain"
], jn = {
  onSelectedEntityChange: "selectedEntityChanged",
  onTrackedEntityChange: "trackedEntityChanged"
}, Nn = [
  "className",
  "id",
  "style",
  "full",
  "containerProps",
  "extend",
  "terrainProvider"
], Er = a({
  name: "Viewer",
  async create(i, { baseLayer: e, terrainProvider: t, ...n }, r) {
    if (!r) return;
    let c;
    I(t) ? c = await t : c = t;
    const l = new Ei(r, {
      ...n,
      terrainProvider: c,
      baseLayer: e ?? void 0
    });
    if (!l) return;
    e === !1 && l.imageryLayers.removeAll(), l && n.extend && (Array.isArray(n.extend) ? n.extend.forEach((u) => {
      l.extend(u, {});
    }) : l.extend(n.extend, {}));
    const m = new ae(l.scene);
    return [l, m];
  },
  destroy(i, e, t, n) {
    n && !n.isDestroyed() && n.destroy(), i.isDestroyed() || i.destroy();
  },
  provide(i, e, t, n) {
    return {
      viewer: i,
      cesiumWidget: i.cesiumWidget,
      dataSourceCollection: i.dataSources,
      entityCollection: i.entities,
      scene: i.scene,
      camera: i.scene.camera,
      imageryLayerCollection: i.scene.globe.imageryLayers,
      primitiveCollection: i.scene.primitives,
      globe: i.scene.globe,
      __$internal: {
        onUpdate: t?.onUpdate,
        imageryLayerWaitingList: []
      },
      [q]: n
    };
  },
  containerProps: ({ id: i, className: e, style: t, full: n, containerProps: r }) => ({
    className: e,
    id: i,
    style: {
      ...n ? {
        position: "absolute",
        bottom: "0",
        left: "0",
        right: "0",
        top: "0"
      } : {},
      ...t
    },
    ...r
  }),
  cesiumProps: Wn,
  cesiumReadonlyProps: Hn,
  cesiumEventProps: jn,
  otherProps: Nn,
  renderContainer: !0,
  useCommonEvent: !0,
  useRootEvent: !0
}), zn = [
  "positions",
  "maximumHeights",
  "minimumHeights",
  "show",
  "fill",
  "material",
  "outline",
  "outlineColor",
  "outlineWidth",
  "granularity",
  "shadows",
  "distanceDisplayCondition"
], xn = {
  onDefinitionChange: "definitionChanged"
}, wr = a({
  name: "WallGraphics",
  create(i, e) {
    if (!i.entity) return;
    const t = new wi(e);
    return i.entity.wall = t, t;
  },
  destroy(i, e) {
    e.entity && (e.entity.wall = void 0);
  },
  cesiumProps: zn,
  cesiumEventProps: xn
});
export {
  or as AmbientOcclusion,
  Xn as Billboard,
  Qn as BillboardCollection,
  eo as BillboardGraphics,
  Xo as BlackAndWhiteStage,
  rr as Bloom,
  sr as BlurStage,
  io as BoxGraphics,
  Qo as BrightnessStage,
  to as Camera,
  no as CameraFlyHome,
  oo as CameraFlyTo,
  so as CameraFlyToBoundingSphere,
  ro as CameraLookAt,
  lo as Cesium3DTilesTerrainProvider,
  ao as Cesium3DTileset,
  co as Cesium3DTilesetGraphics,
  me as CesiumContext,
  uo as CesiumWidget,
  mo as ClassificationPrimitive,
  ho as Clock,
  fo as CloudCollection,
  bi as Consumer,
  yo as CorridorGraphics,
  po as CumulusCloud,
  go as CustomDataSource,
  vo as CylinderGraphics,
  Co as CzmlDataSource,
  ar as DepthOfFieldStage,
  cr as EdgeDetectionStage,
  Po as EllipseGraphics,
  Eo as EllipsoidGraphics,
  wo as Entity,
  So as EntityDescription,
  ae as EventManager,
  bo as Fog,
  ir as Fxaa,
  Do as GeoJsonDataSource,
  Ro as Globe,
  To as Google2DImageryProvider,
  Mo as GroundPolylinePrimitive,
  ko as GroundPrimitive,
  $o as GroundPrimitiveCollection,
  Lo as ImageryLayer,
  _o as ImageryLayerCollection,
  Oo as KmlDataSource,
  Ao as Label,
  Bo as LabelCollection,
  Go as LabelGraphics,
  er as LensFlareStage,
  Io as Model,
  Fo as ModelGraphics,
  Vo as Moon,
  tr as NightVisionStage,
  Uo as ParticleSystem,
  Wo as PathGraphics,
  Ho as PlaneGraphics,
  jo as PointGraphics,
  No as PointPrimitive,
  zo as PointPrimitiveCollection,
  xo as PolygonGraphics,
  qo as Polyline,
  Yo as PolylineCollection,
  Jo as PolylineGraphics,
  Ko as PolylineVolumeGraphics,
  Zo as PostProcessStage,
  nr as PostProcessStageComposite,
  ur as Primitive,
  Kn as Provider,
  dr as RectangleGraphics,
  mr as Scene,
  hr as ScreenSpaceCameraController,
  fr as ScreenSpaceEvent,
  yr as ScreenSpaceEventHandler,
  pr as ShadowMap,
  lr as SilhouetteStage,
  gr as SkyAtmosphere,
  vr as SkyBox,
  Cr as Sun,
  Pr as TimeDynamicPointCloud,
  Er as Viewer,
  wr as WallGraphics,
  se as createCameraOperation,
  a as createCesiumComponent,
  T as createPostProcessStage,
  ue as entries,
  q as eventManagerContextKey,
  de as eventNames,
  ie as includes,
  $i as isDestroyable,
  Li as isDestroyed,
  I as isPromise,
  Mi as pick,
  ki as shallowEquals,
  re as useCesium,
  _i as useCesiumComponent,
  Zn as withCesium
};
